src/rpc/RPCHelpers.cpp

```diff
@@ -405,61 +407,111 @@ getStartHint(ripple::SLE const& sle, ripple::AccountID const& accountID)
     return sle.getFieldU64(ripple::sfOwnerNode);
 }
 
+// traverse account's nfts
+// return Status if error occurs
+// return [nextpage, count of nft already found] if success
 std::variant<Status, AccountCursor>
-traverseOwnedNodes(
+traverseNFTObjects(
     BackendInterface const& backend,
-    ripple::AccountID const& accountID,
     std::uint32_t sequence,
+    ripple::AccountID const& accountID,
+    ripple::uint256 nextPage,
     std::uint32_t limit,
-    std::optional<std::string> jsonCursor,
     boost::asio::yield_context& yield,
     std::function<void(ripple::SLE&&)> atOwnedNode)
 {
-    if (!backend.fetchLedgerObject(ripple::keylet::account(accountID).key, sequence, yield))
-        return Status{RippledError::rpcACT_NOT_FOUND};
+    auto const firstNFTPage = ripple::keylet::nftpage_min(accountID);
+    auto const lastNFTPage = ripple::keylet::nftpage_max(accountID);
 
-    auto const maybeCursor = parseAccountCursor(jsonCursor);
-    if (!maybeCursor)
-        return Status(ripple::rpcINVALID_PARAMS, "Malformed cursor");
+    // check if nextPage is valid
+    if (nextPage != beast::zero and firstNFTPage.key != (nextPage & ~ripple::nft::pageMask))
+        return Status{RippledError::rpcINVALID_PARAMS, "invalid Marker"};
 
-    auto [hexCursor, startHint] = *maybeCursor;
+    // no marker, start from the last page
+    ripple::uint256 currentPage = nextPage == beast::zero ? lastNFTPage.key : nextPage;
 
-    return traverseOwnedNodes(
-        backend,
-        ripple::keylet::ownerDir(accountID),
-        hexCursor,
-        startHint,
-        sequence,
-        limit,
-        jsonCursor,
-        yield,
-        atOwnedNode);
+    // read the current page
+    auto page = backend.fetchLedgerObject(currentPage, sequence, yield);
+
+    if (!page)
+    {
+        if (nextPage == beast::zero)  // no nft objects in lastNFTPage
+            return AccountCursor{beast::zero, 0};
+        else  // marker is in the right range, but still invalid
+            return Status{RippledError::rpcINVALID_PARAMS, "invalid Marker"};
+    }
+
+    // the object exists and the key is in right range, must be nft page
+    ripple::SLE pageSLE{ripple::SLE{ripple::SerialIter{page->data(), page->size()}, currentPage}};
+
+    auto count = 0;
+    // traverse the nft page linked list until the start of the list or reach the limit
+    while (true)
```

Comment:
while(true) is a bit scary :) 
Let's rewrite this as a simple for loop:
```
for(auto count = 0u; count < limit; ++count) { ... break early if previous page is zero ... }
```

Reply:
```
    auto count = 1;
    ripple::uint256 nftPreviousPage = beast::zero;
    // traverse the nft page linked list until the start of the list or reach the limit
    for (; count <= limit; count++)
    {
        nftPreviousPage = pageSLE.getFieldH256(ripple::sfPreviousPageMin);
        atOwnedNode(std::move(pageSLE));

        if (count == limit or nftPreviousPage == beast::zero)
            break;

        page = backend.fetchLedgerObject(nftPreviousPage, sequence, yield);
        pageSLE = ripple::SLE{ripple::SerialIter{page->data(), page->size()}, nftPreviousPage};
    }

    return AccountCursor{nftPreviousPage, count};
```

Not much better. The count check is still inside the for loop. I will keep the while loop, but simplify the inner logic bit.